libbeat/beat/beat.go

```diff
@@ -90,31 +90,52 @@ type Beat struct {
 	Registry *reload.Registry // input, & output registry for configuration manager, should be instantiated in NewBeat
 }
 
-// GenerateUserAgent populates the UserAgent field on the beat.Info struct
-func (beat *Beat) GenerateUserAgent() {
-	// if we're in fleet mode, construct some additional elements for the UA comment field
-	comments := []string{}
-	if beat.Manager != nil && beat.Manager.Enabled() {
-		info := beat.Manager.AgentInfo()
-		if info.ManagedMode == proto.AgentManagedMode_MANAGED {
-			comments = append(comments, "Managed")
-		} else if info.ManagedMode == proto.AgentManagedMode_STANDALONE {
-			comments = append(comments, "Standalone")
-		}
-
-		if info.Unprivileged {
-			comments = append(comments, "Unprivileged")
-		}
+func (beat *Beat) userAgentMode() useragent.AgentManagementMode {
+	if beat.Manager == nil {
+		return useragent.AgentManagementModeUnknown
+	}
+	if !beat.Manager.Enabled() {
+		return useragent.AgentManagementModeStandalone
+	}
+
+	info := beat.Manager.AgentInfo()
+	switch info.ManagedMode {
+	case proto.AgentManagedMode_MANAGED:
+		return useragent.AgentManagementModeManaged
+	case proto.AgentManagedMode_STANDALONE:
+		return useragent.AgentManagementModeUnmanaged
 	}
+	// this is probably not reachable
+	return useragent.AgentManagementModeUnknown
+}
 
-	UserAgentProduct := beat.Info.Beat
-	if UserAgentProduct == "" {
-		UserAgentProduct = "Libbeat"
+func (beat *Beat) userAgentUnprivilegedMode() useragent.AgentUnprivilegedMode {
+	if beat.Manager == nil || !beat.Manager.Enabled() {
+		return useragent.AgentUnprivilegedModeUnknown
 	}
+	if beat.Manager.AgentInfo().Unprivileged {
+		return useragent.AgentUnprivilegedModeUnprivileged
+	}
+	return useragent.AgentUnprivilegedModePrivileged
+}
 
-	finalUserAgent := useragent.UserAgent(UserAgentProduct, version.GetDefaultVersion(),
-		version.Commit(), version.BuildTime().String(), comments...)
-	beat.Info.UserAgent = finalUserAgent
+// generateUserAgent returns the user agent string for the beat.
+func (beat *Beat) generateUserAgent() string {
+	userAgentProduct := "Libbeat"
+	if beat.Info.Beat != "" {
+		userAgentProduct = beat.Info.Beat
+	}
+
+	mode := beat.userAgentMode()
+	unprivileged := beat.userAgentUnprivilegedMode()
+
+	return useragent.UserAgentWithBeatTelemetry(userAgentProduct, version.GetDefaultVersion(),
+		mode, unprivileged)
+}
+
+// GenerateUserAgent populates the UserAgent field on the beat.Info struct
+func (beat *Beat) GenerateUserAgent() {
+	beat.Info.UserAgent = beat.generateUserAgent()
 }
```

Comment:
I don't see the value of having this method just to call another and assign a value. You could have a single method, `GenerateUserAgent` that does what `generateUserAgent` and `GenerateUserAgent` do.

Reply:
good call, this is the result of a bunch of refactoring